src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java

```diff
@@ -112,12 +107,16 @@ public static void success(String k, Object returnValue) {
         State s = State.get();
         StepContext c;
         synchronized (s) {
-            if (!s.contexts.containsKey(k)) {
+            KeyState keyState = s.keyStates.get(k);
+            if (keyState == null || keyState instanceof WillSucceedState || keyState instanceof WillFailState) {
                 LOGGER.info(() -> "Planning to unblock " + k + " as success");
-                s.returnValues.put(k, returnValue);
+                s.keyStates.put(k, new WillSucceedState(returnValue));
                 return;
+            } else if (keyState instanceof FinishedState) {
+                throw new IllegalStateException(k + " already finished");
```

Comment:
This breaks tests. In particular at least `workflow-step-api / org.jenkinsci.plugins.workflow.steps.GeneralNonBlockingStepExecutionTest.stop` and `workflow-cps / org.jenkinsci.plugins.workflow.cps.CpsFlowExecutionTest.stepsAreStoppedWhenCpsVmExecutorServiceHandlesUncaughtException`. I'll take a look.

Reply:
`GeneralNonBlockingStepExecutionTest.stop` I think just has a redundant call to `SemaphoreStep.success` that could be removed.

`CpsFlowExecutionTest.stepsAreStoppedWhenCpsVmExecutorServiceHandlesUncaughtException` is a little unique, basically the test was using `semaphore` as a way to call `StepContext.onSuccess` after a Pipeline died. To preserve that test's behavior without changes, we'd have to make `FinishedState` keep the serialized context around. Probably clearer to update that test to use a custom step instead.